R/combineResults.R

```diff
@@ -0,0 +1,232 @@
+#'
+#' ###############################
+#' ##     Combining results     ##
+#' ###############################
+#'
+#' pred.single <- combineResults(list(pred1, pred2))
+#' pred.clust <- combineResults(list(pred3, pred4))
+#'
+#' @seealso
+#' \code{\link{matchReferences}}, to harmonize labels between reference datasets.
+#' \code{\link{SingleR}}, for generating predictions.
+#'
+#' @importFrom S4Vectors DataFrame metadata metadata<-
+#'
+#' @export
+combineResults <- function(results) {
+    num.features <- sapply(results, nrow)
+    if (abs(max(num.features) - min(num.features)) != 0) {
+        stop("Results DataFrames contain different numbers of cells or clusters.")
+    }
+
+    pred.scores <- list()
+    pred.firstlabels <- list()
+    pred.prunedlabels <- list()
+    pred.tuningscores <- list()
+    feature.names <- list()
+    de.genes <- list()
+    common.genes <- list()
+
+    # Keep track of which scores belong to which results object.
+    pred.score.indices <- list()
+    ind <- 1
+
+    for (i in seq_along(results)) {
+        res <- results[[i]]
+        pred.scores[[i]] <- res$scores
+
+        if (!is.null(res$first.labels)) {
+            pred.firstlabels[[i]] <- res$first.labels
+        } else {
+            pred.firstlabels[[i]] <- NA_character_
+        }
+
+        if (!is.null(res$pruned.labels)) {
+            pred.prunedlabels[[i]] <- res$pruned.labels
+        } else {
+            pred.prunedlabels[[i]] <- NA_character_
+        }
+
+        if (!is.null(res$tuning.scores)) {
+            pred.tuningscores[[i]] <- res$tuning.scores
+        } else {
+            pred.tuningscores[[i]] <- NA_character_
+        }
+
+        feature.names[[i]] <- rownames(res)
+
+        if (!is.null(metadata(res)$de.genes)) {
+            de.genes[[i]] <- metadata(res)$de.genes
+        } else {
+            de.genes[[i]] <- NA_character_
+        }
+
+        common.genes[[i]] <- metadata(res)$common.genes
+        pred.score.indices[[i]] <- seq(ind, (ind + ncol(res$scores) - 1))
+        ind <- ind + ncol(res$scores)
+    }
+
+    all.scores <- do.call(cbind, pred.scores)
+    best.labels <- colnames(all.scores)[max.col(all.scores)]
+
+    # Determine which results DataFrame each score comes from.
+    pred.indices <- sapply(max.col(all.scores), .get_pred_indices, pred.score.indices = pred.score.indices)
+
+    output <- DataFrame(scores = I(all.scores), labels = best.labels)
+
+    if (length(feature.names) > 0) {
+        if (length(unique(feature.names)) == 1) {
+            rownames(output) <- feature.names[[1]]
+        } else {
+            warning("Results Dataframes row names do not match, skipping row name assignment.")
+        }
+    }
+
+    if (length(unique(common.genes)) == 1) {
+        metadata(output)$common.genes <- common.genes[[1]]
+    } else {
+        warning("Results Dataframes common genes do not match, skipping common.genes metadata assignment.")
+    }
+
+    if(!all(is.na(de.genes))) {
```

Comment:
I think you mean to do a `c` here, `de.genes` is a per-label list.

Reply:
I throw the de.genes from each object into their own list (with the completely clear name `de.genes`) in lines 162-166. so I don't think so. This will change once I just check the first `results` element for each field anyway.